swarmcg/simulations/runner.py

```diff
@@ -0,0 +1,131 @@
+    gmx_cmd = f"{ns.gmx_path} {gmx_cmd}"
+    if ns.gmx_args_str != '':
+        gmx_cmd = f"{gmx_cmd} {ns.gmx_args_str}"
+    else:
+        if ns.nb_threads > 0:
+            gmx_cmd = f"{gmx_cmd} -nt {ns.nb_threads}"
+        if len(ns.gpu_id) > 0:
+            gmx_cmd = f"{gmx_cmd} -gpu_id {ns.gpu_id}"
+    if mpi and ns.mpi_tasks > 1:
+        gmx_cmd = f"mpirun -np {ns.mpi_tasks} {ns.gmx_cmd}"
+
+    return gmx_cmd
+
+
+def exec_gmx(gmx_cmd):
+    """Execute gmx cmd and return only exit code"""
+    with subprocess.Popen([gmx_cmd], shell=True, stdout=subprocess.PIPE,
+                          stderr=subprocess.PIPE) as gmx_process:
+        gmx_out = gmx_process.communicate()[1].decode()
+        gmx_process.kill()
+    if gmx_process.returncode != 0:
+        print_stdout_forced(
+            'NON-ZERO EXIT CODE FOR COMMAND:', gmx_cmd, '\n\nCOMMAND OUTPUT:\n\n', gmx_out, '\n\n'
+        )
+    return gmx_process.returncode
+
+
+def cmdline(command):
+    """Execute command and return output"""
+    try:
+        output = subprocess.check_output(command, stderr=subprocess.STDOUT, shell=True).decode()
+        success = True
+    except subprocess.CalledProcessError as e:
+        output = e.output.decode()
+        success = False
+    return success, output
+
+
+class SimulationStep:
+
+    PREP_CMD = "{exec} grompp -c {gro} -f {mdp} -p {top} -n {index} -o {md_input}"
+    MD_CMD = "{exec} mdrun -s {md_input} -nt {n_cpu} -gpu_id {gpu_id} -deffnm {md_output}"
+
+    def __init__(self, sim_setup, step_name):
+        self.sim_setup = sim_setup
+        self.step_name = step_name
+        self._validate_args()
+
+    def _validate_args(self):
+        required_args = ["exec", "gro", "mdp", "top", "index", "md_input", "md_output"]
+        missing_args = ", ".join([i for i in required_args if i not in self.sim_setup.keys()])
+        if missing_args:
+            msg = (
+                "The following arguments are missing: {missing_args}. Please check you input."
+            )
+            raise exceptions.InputArgumentError(msg)
+
+    def _prepare_cmd(self, **kwargs):
+        return SimulationStep.PREP_CMD.format(**{**self.sim_setup, **kwargs})
+
+    def _run_cmd(self, **kwargs):
+        return SimulationStep.MD_CMD.format(**{**self.sim_setup, **kwargs})
+
+    def _run_prep(self, cmd):
+        with subprocess.Popen([cmd], shell=True, stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE) as gmx_process:
+            _ = gmx_process.communicate()[1].decode()
+            gmx_process.kill()
+
+        if gmx_process.returncode == 0:
+            return self
+        else:
+            msg = (
+                f"Gromacs grompp failed at MD {self.step_name} step, see its error message above. " 
+                f"You may also want to check the parameters of the MDP file provided through "
+                f"argument -cg_sim_mdp_mini. If you think this is a bug, please consider opening "
```

Comment:
Here the message may refer to arguments -cg_sim_mdp_mini if the error happened at the minimization step, but it has to be adapted and refer to the correct arguments for also the equilibration and md/production steps.

Reply:
I missed that, I can add another argument to the class for this or via subclassing